spock-specs/src/test/groovy/org/spockframework/docs/interaction/DetachedMockFactoryDocSpec.groovy

```diff
@@ -0,0 +1,149 @@
+package org.spockframework.docs.interaction
+
+import org.spockframework.mock.IDefaultResponse
+import org.spockframework.mock.IMockInvocation
+import org.spockframework.mock.MockUtil
+import org.spockframework.mock.ZeroOrNullResponse
+import spock.lang.Specification
+import spock.lang.Unroll
+import spock.mock.AutoAttach
+import spock.mock.DetachedMockFactory
+
+import static org.spockframework.docs.interaction.DetachedMockFactoryDocSpec.EngineMockCreator.StartMode.*
+
+class DetachedMockFactoryDocSpec extends Specification {
+  // tag::declare-static[]
+  static mockFactory = new DetachedMockFactory()
+  static mockUtil = new MockUtil()
+  // end::declare-static[]
+
+  // tag::attach-manually[]
+  def "Manually attach detached mock"() {
+    given:
+    def manuallyAttachedEngine = mockFactory.Mock(Engine)
+    manuallyAttachedEngine.isStarted() >> true
+    mockUtil.attachMock(manuallyAttachedEngine, this)
+    def car = new Car(engine: manuallyAttachedEngine)
+
+    when:
+    car.drive()
+    then:
+    1 * manuallyAttachedEngine.start()
+    manuallyAttachedEngine.isStarted()
+
+    when:
+    car.park()
+    then:
+    1 * manuallyAttachedEngine.stop()
+    manuallyAttachedEngine.isStarted()
+
+    cleanup:
+    mockUtil.detachMock(manuallyAttachedEngine)
+  }
+  // end::attach-manually[]
+
+  // tag::auto-attach[]
+  @AutoAttach
+  def autoAttachedEngine = mockFactory.Mock(Engine)
+
+  def "Auto-attach detached mock"() {
+    given:
+    autoAttachedEngine.isStarted() >> true
+    def car = new Car(engine: autoAttachedEngine)
+
+    when:
+    car.drive()
+    then:
+    1 * autoAttachedEngine.start()
+    autoAttachedEngine.isStarted()
+
+    when:
+    car.park()
+    then:
+    1 * autoAttachedEngine.stop()
+    autoAttachedEngine.isStarted()
+  }
+  // end::auto-attach[]
+
+  // tag::use-custom-mock-creator[]
+  @Unroll("Engine state #engineStateResponseType")
+  def "Manually attach detached mock with preconfigured engine state"() {
+    given:
+    mockUtil.attachMock(preconfiguredEngine, this)
```

Comment:
As mentioned in the text above, you should show how it behaves without being attached to the specification as well.

Reply:
Um, well, the whole sample spec never overrides the default behaviour and uses the mocks as they are created. So either, I make the example even more verbose by overriding behaviour after attaching the mocks, or I never attach them in the first place. I think, I will do the former and also add some code comments right inside the sample spec, even though it makes the code lengthier.

Reply:
@leonard84, can you explain why after changing the parametrised spec according to your suggestion, it is now failing? Either "ich stehe auf dem Schlauch", as we Germans like to say, and am overlooking something obvious. Or I just uncovered a Spock problem. I would like to have your insight on that. Maybe you want tu run the spec and see what it does. But for the life of me, I do not know why overriding the default behaviour does not work as I would expect it to.

Reply:
This is really strange. I have added two additional tests to isolate the behavior.
`"When block interaction still active after when block without verification"`
`"When block interactions are only active during the when block, if verification present"`

The single verification line `1 * m.start()` in the `then:` block, changes the behavior of the `m.isStarted()` method for the rest of the feature execution.

Reply:
@kriegaex I have added a workaround to the test case.
I guess we should open up a new bug ticket to investigate the problem.

Reply:
@AndreasTu, thanks for investigating and creating the issue. I would, however, prefer not to have the workaround pollute the demo spec and rather wait with the documentation merge after #1759 has been fixed. It would be bad to include a known problem into the docs and explain to users why the feature method looks so strange. We can still reconsider, if it proves to be difficult and time-consuming to fix.

Reply:
Besides, I am enjoying this kind of collaborative work. @leonard84 pushed me to improve test and documentation, I uncovered unexpected Spock behaviour, @AndreasTu created a bug ticket. This is so much better than the "one person, one PR" approach and immediately proved to be valuable. 😊😊😊

Reply:
To unblock this PR from the non trivial problem discussed in #1579, I'd suggest that we simply split this feature into two, i.e., one where the mock is attached in the given block and one where it isn't.

Reply:
@leonard84 Or how about my fix from before?
I reapplied it to this PR. 
This would fix the sample until #1579 gets resolved.

Reply:
IMHO having two features is the better way to showcase this anyway.

Reply:
I splitted the feature into two.